### ISSUE TITLE
Calling multiple ZlibDecompressor in parallel causes adler32 checksum failure

### ISSUE THREAD

**Issue by rocksonhead** [weidai11/cryptopp#596]
msvc 15, 17 and gcc 5.4 64bit compilers
crypto versions 5.6.5 and 6.1
built via cmake on gcc 5.6.5
and vs projects on 5.6.5 and 6.1 - though changing runtime library to /MD
Please see attached project for example code to induce failure on windows.

The failure may not occur on the first run, but running the application several times an exception is thrown in the decompression "Adler32 check error"

The example code uses openmp to parallelize, however it can also be replicated using tbb

[test.zip](https://github.com/weidai11/cryptopp/files/1788987/test.zip)

**Comment by rocksonhead** [comment]
The failure seems to be occurring somewhere around the HuffmanDecoder::Decode
This function seems to be returning 0 lengths.
The message is not printed when performing 1 decompression or in debug mode.

`inline unsigned int HuffmanDecoder::Decode(code_t code, /* out */ value_t &value) const
{
	CRYPTOPP_ASSERT(m_codeToValue.size() > 0);
	LookupEntry &entry = m_cache[code & m_cacheMask];
	code_t normalizedCode = 0;
    if (entry.type != 1)
    {
        normalizedCode = BitReverse(code);
    }

    if (entry.type == 0)
    {
        FillCacheEntry(entry, normalizedCode);
    }

	if (entry.type == 1)
	{
		value = entry.value;
        if (entry.len == 0)
            std::cout << "\nDecode failed" << std::flush;
		return entry.len;
	}
	else
	{
		const CodeInfo &codeInfo = (entry.type == 2)
			? entry.begin[(normalizedCode << m_cacheBits) >> (MAX_CODE_BITS - (entry.len - m_cacheBits))]
			: *(std::upper_bound(entry.begin, entry.end, normalizedCode, CodeLessThan())-1);
		value = codeInfo.value;
		return codeInfo.len;
	}
}`

**Comment by rocksonhead** [comment]
Appreciate it!
Will continue to try and diagnose whats happening in the meantime.
Kind regards

**Comment by rocksonhead** [comment]
It might be associated with the struct LookupEntry.  So far I cannot replicate the problem with an extra param in the struct.....

`	struct LookupEntry
	{
		unsigned int type;
        unsigned int hits = 0;
		union
		{
			value_t value;
			const CodeInfo *begin;
		};
		union
		{
			unsigned int len;
			const CodeInfo *end;
		};
	};`

**Comment by noloader** [comment]
@rocksonhead,

Be careful of the union. Accessing the inactive member is undefined behavior in C++ (though it is fine in C). Also see [Accessing inactive union member and undefined behavior?](https://stackoverflow.com/q/11373203/608639)

**Comment by rocksonhead** [comment]
Scratch my comment about the struct - further testing still produces the error :(
I cannot get my test application to fail on my windows ubuntu subsystem.

**Comment by rocksonhead** [comment]
Came at it with a fresh mind this morning and re-read the code.
My current hypothesis is that a race condition is occurring.....

The Inflators - in this case - are using Singleton HuffmanDecoders
`const HuffmanDecoder& Inflator::GetLiteralDecoder() const
{
	return m_blockType == 1 ? Singleton<HuffmanDecoder, NewFixedLiteralDecoder>().Ref() : m_dynamicLiteralDecoder;
}`

This results in multiple threads accessing the same HuffmanDecoder::Decode simultaneously.
This becomes a problem then both threads are trying to decode the same code.
A reference to the cache entry is retrieved and then filled.

Thread 1 may execute the following
`    if (entry.type == 0)
        FillCacheEntry(entry, normalizedCode);
`
upto the point 
`
	normalizedCode &= m_normalizedCacheMask;
	const CodeInfo &codeInfo = *(std::upper_bound(m_codeToValue.begin(), m_codeToValue.end(), normalizedCode, CodeLessThan())-1);
	if (codeInfo.len <= m_cacheBits)
	{
		entry.type = 1;
`
Thread 2, on the same entry then validates 
`	if (entry.type == 1)
	{
		value = entry.value;
		return entry.len;
	}`

As thread 1 has set the entry type, but not the len, thread 2 evaluates the entry type, and returns an unset entry.len

**Comment by noloader** [comment]
Thanks @rocksonhead.

```
const HuffmanDecoder& Inflator::GetLiteralDecoder() const {
    return m_blockType == 1 ? Singleton<HuffmanDecoder, NewFixedLiteralDecoder>().Ref() : m_dynamicLiteralDecoder;
}
```

Oh yeah, that could do it. 

An example of a similar problem and fix is at [Intialization of base64 and other decoders are not threadsafe](https://github.com/weidai11/cryptopp/issues/376). We need to statically allocate the `FixedLiteralDecoder`. It makes for a bigger executable, but it does not suffer the races.

Good job on tracking it down.

**Comment by noloader** [comment]
Thanks again @rocksonhead.

I remember reading a comment about these classes when Wei implemented them. Wei said something about a bug when using the reference implementation so he re-wrote it in C++. I find both the reference implementation and the existing code to be hairy and I try to avoid both.

My gut feeling is, there are several problems. First, I think all initialization should occur in `HuffmanDecoder::Initialize`. It should not defer until `HuffmanDecoder::Decode`. But I also understand a dynamic code needs to be built on the fly.

Second, I think we might need to stop building the static codes at runtime. That strategy has gotten us in trouble in the past.

Third, if we look at `Inflator`:

```
class Inflator
{
    ...
    HuffmanDecoder m_dynamicLiteralDecoder, m_dynamicDistanceDecoder;
};
```

The `Inflator class violates the thread safety we promise by using the `Singleton` to share does between instances. I think the fix is to re-write these two methods:

```
const HuffmanDecoder& Inflator::GetLiteralDecoder() const
{
    return m_blockType == 1 ? Singleton<HuffmanDecoder, NewFixedLiteralDecoder>().Ref() : m_dynamicLiteralDecoder;
}

const HuffmanDecoder& Inflator::GetDistanceDecoder() const
{
    return m_blockType == 1 ? Singleton<HuffmanDecoder, NewFixedDistanceDecoder>().Ref() : m_dynamicDistanceDecoder;
}
```

I'm going to need more time to study this more. Also, let's ping @denisbider, @mouse07410 and @MarcelRaad. They are also maintainers, and they jump in on the non-trivial stuff.

**Comment by noloader** [comment]
@rocksonhead,

Also, if you need a quick and dirty work around, I think it could be similar to the following.

The rub is, it only works reliably for C++11 on Unix and Linux. It does not work for C++03 and below, and it does not work on most Windows machines.

```
const HuffmanDecoder& Inflator::GetLiteralDecoder() const
{
    static HuffmanDecoder decoder;
    static std::once_flag flag;
    
    std::call_once(flag, [](){
        decoder = m_blockType == 1 ? Singleton<HuffmanDecoder, NewFixedLiteralDecoder>().Ref() : m_dynamicLiteralDecoder;
    });
    
    return decoder;
}
```

**Comment by noloader** [comment]
@rocksonhead,

Also, if you need a quick and dirty work around, I think it could be similar to the following.

The rub is, it only works reliably for C++11 on Unix and Linux. It does not work for C++03 and below, and it does not work on most Windows machines.

```
const HuffmanDecoder& Inflator::GetLiteralDecoder() const
{
    static HuffmanDecoder s_FixedLiteralDecoder;
    static std::once_flag s_flag;
    
    std::call_once(flag, [](){
		s_FixedLiteralDecoder = Singleton<HuffmanDecoder, NewFixedLiteralDecoder>().Ref();
    });
	
	return m_blockType == 1 ? s_FixedLiteralDecoder : m_dynamicLiteralDecoder;
}
```

**Comment by denisbider** [comment]
Aw, yikes. This looks pretty bad.

After looking at this for perhaps 30 minutes, I see two issues, one in `Singleton` and one in `Inflator` / `HuffmanDecoder`.

The `Singleton` issue is that the legacy implementation of `Singleton::Ref` in misc.h - the version that's not C++11 - appears to have a race condition. If I'm seeing it right, this can allow the singleton to be initialized multiple times by separate threads. All but one of the created objects will have their pointers lost and will never be destroyed. But worse, the thread that created each object will return the object it created. Therefore, `Singleton::Ref` is not guaranteed to return the same object consistently. In a race condition on first access, different threads can get different objects, but only one is the "true" singleton that won the race. This can lead to inconsistency in application state if threads rely on all getting the same object.

This issue can be relatively straightforwardly fixed. The `Singleton` initialization logic is broken, but it's just a matter of replacing it with more robust logic (that does not have to be any less efficient).

The `Inflator` / `HuffmanDecoder` issue though... Oh my. So the situation we have here is that when `Inflator::m_blockType` equals 1 - a situation controlled by a potential attacker - multiple threads will call `Decode` on the same `HuffmanDecoder` instance previously initialized via `Singleton::Ref`. (Or sometimes perhaps different instances, due to the previous bug in `Singleton::Ref`, hehe.)

The issue here is that the literal `HuffmanDecoder` objects are not immutable, and their initialization doesn't fully happen in `NewFixedLiteralDecoder` or `NewFixedDistanceDecoder`. The `m_cache` is not initialized, and its initialization is deferred until some call to `Decode` occurs that needs a particular `m_cache` entry.

@noloader, this means your above workaround won't work. If I'm seeing this correctly, it may be that the only safe way that currently exists for a multithreaded application to use `Inflator` is if the main thread creates an `Inflator` instance and feeds to it contrived data that will initialize all `m_cache` entries in both the fixed literal encoder, and the fixed distance encoder, *before* any other threads are created.

I see the following ways to fix this in zinflate.cpp:

1. High-maintenance, high-risk, but efficient solution. In `NewFixedLiteralDecoder` and `NewFixedDistanceDecoder`, after the call to `pDecoder->Initialize`, insert a bunch of code that can directly invoke this function:

       inline unsigned int HuffmanDecoder::Decode(code_t code, /* out */ value_t &value) const

   ... with parameters crafted to initialize all `m_cache` entries. This is high-maintenance because if there's any change to `HuffmanDecoder` internal state, this initialization code needs to be updated as well, and there will be no obvious indication that it needs to be updated. It is high-risk because there may be other delayed initialization that can happen besides `m_cache`, and there may be a bug in the initialization code that fails to properly initialize all of `m_cache`.

   However, it is efficient because multiple threads will continue to share the same fixed literal decoder and distance decoder instances.

2. The low-maintenance, low-risk, less efficient, easy to implement solution: Simply make every instance of `Inflator` use its own fixed literal decoder and fixed distance decoder. That is, create new members `m_fixedLiteralDecoder` and `m_fixedDistanceEncoder`, and just use those. Fixed!

3. The low-maintenance, low-risk, medium efficient, harder to implement solution: A variant on 2, but instead of using separate fixed decoders for each `Inflator` instance, use thread local storage to have separate fixed decoder instances for each thread. Problematic because there are constraints on the availability of thread local storage.

I would go with option 2 to fix `Inflator` - optimize safety and ease of maintenance at the expense of efficiency. I would also fix the legacy version of `Singleton::Ref`.

**Comment by denisbider** [comment]
@rocksonhead: Yes, I think the solution you described in [this comment](https://github.com/weidai11/cryptopp/issues/596#issuecomment-372715474) is equivalent to my suggestion 2 in my previous comment, and may perhaps be the most sensible way to fix this in the library, as well.

**Comment by denisbider** [comment]
@noloader, @rocksonhead:

After studying this a bit more, I am wondering if it's perhaps premature to assume that Wei did not intend `HuffmanDecoder` to be thread-safe. Given that `m_cache` is declared mutable:

    mutable std::vector<LookupEntry, AllocatorWithCleanup<LookupEntry> > m_cache;

... it seems clear the author intended this to be a special variable which `const` member methods could safely modify. This is consistent with that `Singleton::Ref` returns a `const` reference.

In this light, perhaps all that is wrong is that `FillCacheEntry` does not actually initialize `m_cache` safely. My understanding is that the function is already idempotent. So perhaps all that's needed is to make it like this:

    void HuffmanDecoder::FillCacheEntry(LookupEntry &entry, code_t normalizedCode) const
    {
        normalizedCode &= m_normalizedCacheMask;
        const CodeInfo &codeInfo = *(std::upper_bound(m_codeToValue.begin(), m_codeToValue.end(), normalizedCode, CodeLessThan())-1);
        if (codeInfo.len <= m_cacheBits)
        {
            entry.value = codeInfo.value;
            entry.len = codeInfo.len;
            MEMORY_BARRIER();
            entry.type = 1;
        }
        else
        {
            entry.begin = &codeInfo;
            const CodeInfo *last = & *(std::upper_bound(m_codeToValue.begin(), m_codeToValue.end(), normalizedCode + ~m_normalizedCacheMask, CodeLessThan())-1);
            if (codeInfo.len == last->len)
            {
                entry.len = codeInfo.len;
                MEMORY_BARRIER();
                entry.type = 2;
            }
            else
            {
                entry.end = last+1;
                MEMORY_BARRIER();
                entry.type = 3;
            }
        }
    }

That being said, though - I'm not sure whether this means that `HuffmanDecoder::Decode` would need a `MEMORY_BARRIER()` after it reads `entry.type` and before it reads `entry.value` and `entry.len`. And if `Decode` needs a memory barrier, I don't know how much less efficient `Decode` would become.

If `Decode` needs a memory barrier, and this would make it less efficient, then that probably defeats the purpose of using the singletons to begin with.

**Comment by denisbider** [comment]
OK, I have educated myself on MEMORY_BARRIER(). The way this is defined in misc.h, it's a compiler-only memory barrier. It emits no instructions, it just prevents compiler reordering of loads and stores.

I believe this means MEMORY_BARRIER() is close to zero-cost, apart from any avoided optimizations. However, it may not have effect on non-x86/64 platforms, e.g. on ARM, which do not provide the same assurances as x86/64 with regard to the order of memory access.

**Comment by rocksonhead** [comment]
I have updated the FillCacheEntry to use the MEMORY_BARRIER() and over several thousand iterations of my test case have yet to hit the race.

**Comment by noloader** [comment]
Thanks @denisbider and @rocksonhead,

I think I favor Denis' solution number (2) based on @rocksonhead suggestions for the long term fix. I started testing the code below yesterday but got pulled away. The code below uses a pointer to the fixed length gear, and allocates it per-instance if needed. To strip it down to its essence:

```
class Inflator
{
    ...
    HuffmanDecoder m_dynamicLiteralDecoder, m_dynamicDistanceDecoder;
    member_ptr<HuffmanDecoder> m_fixedLiteralDecoder, m_fixedDistanceDecoder;
};

 const HuffmanDecoder& Inflator::GetLiteralDecoder() const
 {
    if (m_blockType == 1)
    {
        if (m_fixedLiteralDecoder.get() == NULLPTR)
            m_fixedLiteralDecoder.reset(new FixedLiteralDecoder);
        return *m_fixedLiteralDecoder.get();
    }
    return m_dynamicLiteralDecoder;
 }
```

I'm wondering about the memory barrier for two reasons. First, it may not be available everywhere even though it is available on the common platforms. Second, we started using it for a similar case way back when (the `static bool s_initialized` single flag case). I'm wondering if it will hold up to three variables.

Based on @rocksonhead description of what he is observing, it feels more like we want something more like the following (in an awful pseudo code):

```
lock {
        entry.value = codeInfo.value;
        entry.len = codeInfo.len;
        entry.type = 1;
}
```

Above, we are guarding the whole struct an not a single entry. So it seems like we want a `CRITICAL_SECTION` on Windows or a condition variable on pthreads. That begs the question, what do we do on other platforms without Windows synchronization or pthreads.

I'm liking @denisbider case (2) more and more.

-----

```
$ cat huffman.diff 
diff --git a/zinflate.cpp b/zinflate.cpp
index 9a41462..57bb02e 100644
--- a/zinflate.cpp
+++ b/zinflate.cpp
@@ -612,41 +612,57 @@ void Inflator::FlushOutput()
 	}
 }
 
-struct NewFixedLiteralDecoder
+struct FixedLiteralDecoder : public HuffmanDecoder
 {
-	HuffmanDecoder * operator()() const
+	FixedLiteralDecoder()
 	{
-		unsigned int codeLengths[288];
 		std::fill(codeLengths + 0, codeLengths + 144, 8);
 		std::fill(codeLengths + 144, codeLengths + 256, 9);
 		std::fill(codeLengths + 256, codeLengths + 280, 7);
 		std::fill(codeLengths + 280, codeLengths + 288, 8);
-		member_ptr<HuffmanDecoder> pDecoder(new HuffmanDecoder);
-		pDecoder->Initialize(codeLengths, 288);
-		return pDecoder.release();
+		HuffmanDecoder::Initialize(codeLengths, 288);
 	}
+
+	virtual ~FixedLiteralDecoder() { }
+
+	unsigned int codeLengths[288];
 };
 
-struct NewFixedDistanceDecoder
+struct FixedDistanceDecoder : public HuffmanDecoder
 {
-	HuffmanDecoder * operator()() const
+	FixedDistanceDecoder()
 	{
-		unsigned int codeLengths[32];
 		std::fill(codeLengths + 0, codeLengths + 32, 5);
-		member_ptr<HuffmanDecoder> pDecoder(new HuffmanDecoder);
-		pDecoder->Initialize(codeLengths, 32);
-		return pDecoder.release();
+		HuffmanDecoder::Initialize(codeLengths, 32);
 	}
+
+	virtual ~FixedDistanceDecoder() { }
+
+	unsigned int codeLengths[32];
 };
 
 const HuffmanDecoder& Inflator::GetLiteralDecoder() const
 {
-	return m_blockType == 1 ? Singleton<HuffmanDecoder, NewFixedLiteralDecoder>().Ref() : m_dynamicLiteralDecoder;
+	if (m_blockType == 1)
+	{
+		if (m_fixedLiteralDecoder.get() == NULLPTR)
+			m_fixedLiteralDecoder.reset(new FixedLiteralDecoder);
+		return *m_fixedLiteralDecoder.get();
+	}
+
+	return m_dynamicLiteralDecoder;
 }
 
 const HuffmanDecoder& Inflator::GetDistanceDecoder() const
 {
-	return m_blockType == 1 ? Singleton<HuffmanDecoder, NewFixedDistanceDecoder>().Ref() : m_dynamicDistanceDecoder;
+	if (m_blockType == 1)
+	{
+		if (m_fixedDistanceDecoder.get() == NULLPTR)
+			m_fixedDistanceDecoder.reset(new FixedDistanceDecoder);
+		return *m_fixedDistanceDecoder.get();
+	}
+
+	return m_dynamicDistanceDecoder;
 }
 
 NAMESPACE_END
diff --git a/zinflate.h b/zinflate.h
index e352aff..2287997 100644
--- a/zinflate.h
+++ b/zinflate.h
@@ -148,6 +148,7 @@ private:
 	NextDecode m_nextDecode;
 	unsigned int m_literal, m_distance;	// for LENGTH_BITS or DISTANCE_BITS
 	HuffmanDecoder m_dynamicLiteralDecoder, m_dynamicDistanceDecoder;
+	member_ptr<HuffmanDecoder> m_fixedLiteralDecoder, m_fixedDistanceDecoder;
 	LowFirstBitReader m_reader;
 	SecByteBlock m_window;
 	size_t m_current, m_lastFlush;
```

**Comment by noloader** [comment]
@rocksonhead, @denisbider,

I got a better look at @rocksonhead [fix in comment 372715474](https://github.com/weidai11/cryptopp/issues/596#issuecomment-372715474). The formatting made a word of difference (my ISP seems to be caching some sites, so I don't get to see the updates even after refreshing the page). So it looks like we all came to the same conclusion.

@denisbider, are you OK with @rocksonhead proposed solution?

If so, @rocksonhead, can you make the PR? You worked hard for it. I think you deserve it.

**Comment by rocksonhead** [comment]
One last question ;) 
are the function declarations 
`	static const HuffmanDecoder *FixedLiteralDecoder();
	static const HuffmanDecoder *FixedDistanceDecoder();`
in zinflate.h still required?
VS tells me that there is no function body associated with them... and I can not find them either!

**Comment by noloader** [comment]
@rocksonhead,

The change in [comment 372715474](https://github.com/weidai11/cryptopp/issues/596#issuecomment-372715474) tested good on x86_64, Aarch64 and PPC64-be.

I'm OK with a PR if @denisbider is OK with it.

@denisbider - Should `NewFixedLiteralDecoder` and `NewFixedDistanceDecoder` be made a freestanding function to avoid this? Something looks odd about it.

```
m_literalDecoder.reset(NewFixedLiteralDecoder()());
```

**Comment by rocksonhead** [comment]
Something along the lines ....

`const HuffmanDecoder& Inflator::GetLiteralDecoder() const
{
    if (m_blockType == 1)
    {
        if (m_literalDecoder.get() == NULLPTR)
            CreateFixedLiteralDecoder();
        return *m_literalDecoder;
    }
    else
    {
        return m_dynamicDistanceDecoder;
    }
}`

**Comment by denisbider** [comment]
@noloader: I resolved this in our copy of Crypto++ (which is used on x86/x64 architectures only) by reordering memory writes in `FillCacheEntry` and adding compiler memory barriers in `FillCacheEntry` and `Decode`. There was no measurable impact on performance.

**Comment by noloader** [comment]
Cleared at [Commit b0f717059556](https://github.com/weidai11/cryptopp/commit/b0f717059556).
